### ISSUE TITLE
NEW PROPS- Text Input (Add On)

### ISSUE THREAD

**Issue by jasoncypret** [powerhome/playbook#1190]
##### Kit Background & Use Cases
This will be used to help identify how the text input will be used. These are not buttons.

##### Screens & Design References
![image](https://user-images.githubusercontent.com/863031/98140250-d4d54980-1e8a-11eb-8597-c74d1cf738d2.png)

##### Kit Specifications 
* The add on should be able to take text (example: "%", "$") Limit to character
* The add on should be able to take an icon name
* Should be able to set alignment
* Should be able to set border

##### Example Props (discuss with engineer)

**HASH**
* add_on { icon: "", border: "", text: "", alignment: "" }

**NON HASH**
* add_on_icon: "icon_name"
* add_on_border: (default false)
* add_on_text: "string"
* add_on_alignment: "enum"

**Comment by that-jill** [comment]
@jasoncypret a few questions:

1. Can you tell me a little more about what add_on_text is and how it differs from the placeholder already in `pb_text_input`? Where would it go in relation to the icon? It looks like the limit is a single character? Is the idea here to make the "icon" either a fa-icon or a single character?
2. What should the default be for alignment? And the options should just be LEFT and RIGHT, correct?
3. What should be default be for icon_name?
4. What should the default be for add_on_border? (false, maybe? Or true?)

**Comment by that-jill** [comment]
Updates upon further discussion at dailies:

* we've decided to forgo the text option for now (both the cited examples are also fa-icons)
* if the user decides they want to use an add-on, they'll have to pass in a font-awesome icon
* based upon whether the user passes in the [:icon] argument, we will optionally set the alignment and border, though these will have defaults
* thus, the user can either pass in no add_on hash, or an add_on hash with 1-3 keys.

**Comment by that-jill** [comment]
@xjunior I believe I'm supposed to ask you questions about how to implement this on the Rails side, since there seems to be no current precedent in playbook and I'd be setting one. 

TLDR: we want a new prop on the Text Input: `add_on`, which is a hash. The user/developer has the option of not using this add_on at all, in which case, it'd be an empty hash. However, if the developer opts to use an `add_on`, they are required to pass in an `icon` key to the hash. They are also optionally allowed to pass in two other keys: `alignment` (an enum which defaults to `'right'`) and `border` (a boolean which defaults to `true`). I don't see a pattern for building this kind of `Playbook::Props::Hash` inside of playbook -- what is the intent? Should I add a new type to `Playbook::Props` that really only belongs to this specific use case? Should I put a new prop on the TextInput class?

**Comment by thestephenmarshall** [comment]
@xjunior I mentioned this during dailies to @that-jill yesterday and was brainstorming.. My initial thoughts were to use `props/hash.rb` as a superclass and compose a specific prop type in the kit itself e.g. `Playbook::Props::TextInputAddon`, but then I thought "perhaps we could implement something more abstract that could have at least some shallow hash key validation with the same approach and keep it in the `props` dir instead... then that could be the superclass for composing said Playbook::Props::TextInputAddon". 

Curious what you think..
